### PR TITLE
Add Biometric Imprint

### PR DESCRIPTION
**Pull request by KarimNahas**

### CODE REVIEW

File: app/controllers/cattle_controller.rb [hunk 1]
```diff
@@ -6,12 +6,18 @@ def index
   end
 
   def new
-    cattle = current_user.cattle.create(params.permit(Cattle.column_names))
-    if cattle.valid?
-      cattle.save
-      render json: { cattle: cattle }, status: :created
+    if params[:muzzle_image]
+      cattle = current_user.cattle.create(params.permit(Cattle.column_names))
+      cattle.imprint(params[:muzzle_image])
```
Comment (FreddieLindsey) [inline app/controllers/cattle_controller.rb]: Actually, can we call this `push_imprint`

File: app/controllers/cattle_controller.rb [hunk 2]
```diff
@@ -6,12 +6,18 @@ def index
...
+      cattle.add_image(params[:body_image]) if params[:body_image]
```
Comment (FreddieLindsey) [inline app/controllers/cattle_controller.rb]: Let's call this `push_profile` or whatever convention we use above

File: app/controllers/image_controller.rb [hunk 3]
```diff
@@ -27,17 +15,8 @@ def upload
...
-      image = cattle.image.create(image_uri: 'temporary')
-      image_uri = "cattle/#{current_user.id}/#{cattle.id}/#{image.id}-image-original"
-      $s3.put_object(
-        acl: 'private',
-        body: params[:data],
-        bucket: 'cowhub-production-images',
-        key: image_uri
-      )
-      image.image_uri = image_uri
+      image = cattle.add_image(params[:data])
```
Comment (FreddieLindsey) [inline app/controllers/image_controller.rb]: Same as above, `push_profile`? or is it `push_imprint`?

File: app/controllers/image_controller.rb [hunk 4]
```diff
@@ -27,17 +15,8 @@ def upload
...
       if image.valid?
-        image.save
```
Comment (FreddieLindsey) [inline app/controllers/image_controller.rb:40]: Why have you removed this?

File: app/models/cattle.rb [hunk 5]
```diff
@@ -24,6 +25,48 @@ def generate_tag
     "#{country_code}#{herdmark}#{check_digit}#{format('%05d', individual_number)}"
   end
 
+  def imprint(data)
```
Comment (FreddieLindsey) [inline app/models/cattle.rb]: So this will be `push_imprint`

File: app/models/cattle.rb [hunk 6]
```diff
@@ -24,6 +25,48 @@ def generate_tag
...
+      body: data,
+      bucket: 'cowhub-production-images',
+      key: image_uri
+    )
+    imprint.image_uri = image_uri
+    imprint.save
+    imprint
+  end
+
+  def add_image(data)
```
Comment (FreddieLindsey) [inline app/models/cattle.rb]: `push_profile`

File: app/models/cattle.rb [hunk 7]
```diff
@@ -24,6 +25,48 @@ def generate_tag
...
+      body: data,
+      bucket: 'cowhub-production-images',
+      key: image_uri
+    )
+    profile.image_uri = image_uri
+    profile.save
+    profile
+  end
+
+  def images
```
Comment (FreddieLindsey) [inline app/models/cattle.rb]: You need to split this into two functions, `biometric_imprint_images` and `profile_images`

File: app/models/biometric_imprint.rb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+class BiometricImprint < ApplicationRecord
+  belongs_to :cattle
+
+  validates :cattle, presence: true
```
Comment (FreddieLindsey) [inline app/models/biometric_imprint.rb]: I don't think you need this, it's implied by the above

File: app/models/profile_picture.rb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+class ProfilePicture < ApplicationRecord
+  belongs_to :cattle
+
+  validates :cattle, presence: true
```
Comment (FreddieLindsey) [inline app/models/profile_picture.rb]: I don't think you need this, it's implied by the above

File: app/models/profile_picture.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class ProfilePicture < ApplicationRecord
```
Comment (FreddieLindsey) [inline app/models/profile_picture.rb]: `ProfileImage`, keep it consistent

File: app/models/biometric_imprint.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+class BiometricImprint < ApplicationRecord
```
Comment (FreddieLindsey) [inline app/models/biometric_imprint.rb]: `BiometricImprintImage` please

File: circle.yml [hunk 12]
```diff
@@ -25,7 +25,8 @@ dependencies:
 
 test:
   override:
-    - bundle exec rake test
```
Comment (FreddieLindsey) [inline circle.yml:28]: This change is actually already on master I thought.

File: spec/controllers/cattle_controller_spec.rb [hunk 13]
```diff
@@ -33,15 +33,24 @@
 
   describe 'POST #new' do
     it 'post registration of unregistered tags' do
-      post :new, params: FactoryGirl.attributes_for(:cattle)
+      params = FactoryGirl.attributes_for(:cattle)
```
Comment (FreddieLindsey) [inline spec/controllers/cattle_controller_spec.rb:36]: If you have to have the muzzle_image attribute, make it default

File: spec/controllers/cattle_controller_spec.rb [hunk 14]
```diff
@@ -33,15 +33,24 @@
...
     it 'post registration of unregistered tags with extra detail' do
-      post :new, params: FactoryGirl.attributes_for(:cattle_extended)
+      params = FactoryGirl.attributes_for(:cattle_extended)
+      params[:muzzle_image] = SecureRandom.base64
+      post :new, params: params
       expect(response).to have_http_status(:created)
     end
 
+    it 'post registration of unregistered tags' do
+      post :new, params: FactoryGirl.attributes_for(:cattle)
```
Comment (FreddieLindsey) [inline spec/controllers/cattle_controller_spec.rb:50]: We should be deleting required attributes here. Naming of the test doesn't seem right either.

File: spec/controllers/image_controller_spec.rb [hunk 15]
```diff
@@ -4,7 +4,7 @@
   before(:all) do
     @user = FactoryGirl.create(:user)
     @cattle = FactoryGirl.create(:cattle, user_id: @user.id)
-    FactoryGirl.create_list(:image, 20, cattle_id: @cattle.id)
+    FactoryGirl.create_list(:profile_picture, 20, cattle_id: @cattle.id)
```
Comment (FreddieLindsey) [inline spec/controllers/image_controller_spec.rb]: Correspond to `ProfileImage` change above

**Review by FreddieLindsey — CHANGES_REQUESTED** [review]
Please make modifications as written.

File: app/controllers/image_controller.rb (hunk 4 above)
Comment (KarimNahas) [reply app/controllers/image_controller.rb:40]: because I do it in the add  method

File: config/initializers/inflections.rb [hunk 16]
```diff
@@ -1,3 +1,4 @@
 ActiveSupport::Inflector.inflections do |inflect|
   inflect.plural 'cattle', 'cattle'
+  inflect.plural 'imprint_image', 'imprints_images'
```
Comment (FreddieLindsey) [inline config/initializers/inflections.rb:3]: Don't need this??

File: spec/controllers/cattle_controller_spec.rb (hunk 13 above)
Comment (FreddieLindsey) [reply spec/controllers/cattle_controller_spec.rb:36]: Resolved

**Review by FreddieLindsey — APPROVED** [review]

**Pull request merged**